__tests__/ExpensiMark-HTML-test.js

```diff
@@ -399,7 +402,10 @@ test('Test url replacements', () => {
         + '<a href="https://google.com/maps/place/The+Flying&#x27;+Saucer/@42.4043314,-86.2742418,15z/data=!4m5!3m4!1s0x0:0xe28f6108670216bc!8m2!3d42.4043316!4d-86.2742121" target="_blank" rel="noreferrer noopener">google.com/maps/place/The+Flying&#x27;+Saucer/@42.4043314,-86.2742418,15z/data=!4m5!3m4!1s0x0:0xe28f6108670216bc!8m2!3d42.4043316!4d-86.2742121</a> '
         + '<a href="https://google.com/maps/place/%E9%9D%92%E5%B3%B6%E9%80%A3%E7%B5%A1%E8%88%B9%E4%B9%97%E5%A0%B4/@33.7363156,132.4877213,17.78z/data=!4m5!3m4!1s0x3545615c8c65bf7f:0xb89272c1a705a33f!8m2!3d33.7366776!4d132.4878843" target="_blank" rel="noreferrer noopener">google.com/maps/place/%E9%9D%92%E5%B3%B6%E9%80%A3%E7%B5%A1%E8%88%B9%E4%B9%97%E5%A0%B4/@33.7363156,132.4877213,17.78z/data=!4m5!3m4!1s0x3545615c8c65bf7f:0xb89272c1a705a33f!8m2!3d33.7366776!4d132.4878843</a> '
-        + '<a href="https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422" target="_blank" rel="noreferrer noopener">https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422</a>';
+        + '<a href="https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422" target="_blank" rel="noreferrer noopener">https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422</a> '
+        + '<a href="https://www.facebook.com/hashtag/__main/?__eep__=6" target="_blank" rel="noreferrer noopener">https://www.facebook.com/hashtag/__main/?__eep__=6</a> '
+        + '<a href="https://example.com/~username/foo~bar.txt" target="_blank" rel="noreferrer noopener">https://example.com/~username/foo~bar.txt</a> '
+        + '<a href="http://example.com/foo/*/bar/*/test.txt" target="_blank" rel="noreferrer noopener">http://example.com/foo/*/bar/*/test.txt</a>';
```

Comment:
Hmm can you help me understand why we expect `http://example.com/foo/*/bar/*/test.txt` to be identified as and converted to a link? I'm reading [this article](https://perishablepress.com/stop-using-unsafe-characters-in-urls/#:~:text=These%20characters%20are%20%7B%20%2C%20%7D%20%2C,be%20encoded%20within%20a%20URL.) which references the [official spec](https://www.rfc-editor.org/rfc/rfc3986) and my takeaway is that `*` is a reserved character and would need to be URL-encoded in order to be valid, so I'm not sure we should have this as a test case (but I could also be misunderstanding something).

Reply:
Hi @amyevans , that's a great question! 

The article you linked says that

> 1 Reserved characters only need encoded when not used for their defined, reserved purposes.

When I tried to add the case of `http://example.com/foo/*/bar/*/test.txt`, I also found that the asterisk character is not alway valid in a url but has to be used in its defined and reserved purpose way. This test case url `http://example.com/foo/*/bar/*/test.txt` should follow the defined purpose. I checked it with the [web url regex](https://github.com/Expensify/expensify-common/blob/be625f55058111793bbcfe3f64788ea0a8f9d705/lib/Url.js#L8) provided by this repo, see

<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/232636027-92c6b0e0-f9bb-41ff-ad80-c956bfdc552c.png">

You may test more cases about link includes the asterisk characters and below the web url regex could be helpful 
```js
/((https?:\/\/)?((?:www\.)?[-a-z0-9]+?\.)+(?:COM|ORG)(?:\:\d{2,4}|\b|(?=_))(?:(?:[.,=(+$!*]|&(?:amp|quot|#x27);)?\/(?:[-\w$@.+!*:(),=%~]|&(?:amp|quot|#x27);)*(?:[-\w~@:%)]|&(?:amp|quot|#x27);)|\/)*(?:(?:\?(?:[-\w$@.+!*()\/,=%{}:;\[\]\|_]|&(?:amp|quot|#x27);)*)?|(?:#(?:[-\w$@.+!*()[\],=%;\/:~]|&(?:amp|quot|#x27);)*)?)*)/
```

Reply:
Makes sense, thank you both for the additional discussion!